finagle-redis/src/main/scala/com/twitter/finagle/Redis.scala

```diff
@@ -63,4 +66,48 @@ object Redis extends Client[Command, Reply] {
 
   def newService(dest: Name, label: String): Service[Command, Reply] =
     client.newService(dest, label)
+
+  object Subscribe extends com.twitter.finagle.Client[SubscribeCommand, Unit] {
+
+    object SubscribeClient {
+      /**
+       * Default stack parameters used for redis client.
+       */
+      val defaultParams: Stack.Params = StackClient.defaultParams +
+        param.ProtocolLibrary("redis.subscribe")
+
+      /**
+       * A default client stack which supports the pipelined redis client.
+       */
+      def newStack: Stack[ServiceFactory[SubscribeCommand, Unit]] = StackClient.newStack
+        .replace(DefaultPool.Role, SingletonPool.module[SubscribeCommand, Unit])
+    }
+
+    case class SubscribeClient(
```

Comment:
Yes, transactions are unrelated to subscription, except in that when you're in a transaction, you can't use that connection for anything else, the same way that you can't use a subscription connection for anything else.

If we changed the connection pool API slightly, so that it was something like:

```scala
trait ReservingPool[Req, Rep] {
  // nobody else will use this connection until it's marked as no longer needed (ie closed)
  def reserve(): Service[Req, Rep]

  // may be reused
  def get(): Service[Req, Rep]
}
```

Then we could use it with redis and be guaranteed of not violating the protocol.

1.  When you say we are not replied when we issue an unsubscribe command, but we will receive an acknowledgment, do you mean the TCP ack, or does redis also provide an acknowledgment?  This seems pretty doable–we can have something which doesn't mark the reserved connection as "closed" until after we've received the acknowledgment.

2.  Yeah, we have tools like this for memcached (see the sharded multiget) but I'm a little hesitant to start building sharding into our clients.  It might make sense to make sharding helpers at first so that it's easy for folks to make sharders that work for their use case.  In the long run, it would be nice to have a toolkit that we can hook into any protocol that we need to shard manually (redis, mysql, memcached) but for now it probably makes sense to just make little custom helpers.

Another thing to consider is that a lot of the tooling finagle has is around naming, and balancing between hosts.  This isn't something that we can do in something like redis, because we only rarely consider different hosts to be the same (ie in redis cluster mode).

It might make sense to see what we can do to simplify the redis client by stripping the parts we don't need.

we should be able to remove:
both of the FactoryToService
BindingFactory
AddrMetadataExtraction
Loadbalancing
FailFast
FailureAccrual
DtabStatsFilter

Anyway, most of these are out of scope.  We should just set them as TODOs for future work.